apis/duck/v1/source_types.go

```diff
@@ -170,3 +171,50 @@ type SourceList struct {
 
 	Items []Source `json:"items"`
 }
+
+func (s *Source) Validate(ctx context.Context) *apis.FieldError {
+	if s == nil {
+		return nil
+	}
+	return ValidateSource(ctx, *s).ViaField(apis.CurrentField)
+}
+
+// ValidateSource validates Source.
+func ValidateSource(ctx context.Context, src Source) *apis.FieldError {
+	return src.Spec.Validate(ctx).ViaField("spec")
+}
+
+func (s *SourceSpec) Validate(ctx context.Context) *apis.FieldError {
+	return s.CloudEventOverrides.Validate(ctx).ViaField("ceOverrides")
+}
+
+func (ceOverrides *CloudEventOverrides) Validate(ctx context.Context) *apis.FieldError {
+	for key := range ceOverrides.Extensions {
+		if err := validateExtensionName(key); err != nil {
+			return err.ViaField("extensions")
+		}
+	}
+	return nil
+}
+
+func validateExtensionName(key string) *apis.FieldError {
+	if len(key) < 1 {
+		return apis.ErrInvalidKeyName(
+			key,
+			"",
+			"CloudEvents attribute names MUST NOT be empty",
+		)
+	}
+
+	for _, c := range key {
+		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
```

Comment:
only lower-case letters or digits are allowed.

Reply:
Yup, should i comment that? this is the exact condition that the cloudevents go sdk uses

Reply:
really? hmm maybe a bug in the SDK?

Reply:
Here is the go-sdk line: https://github.com/cloudevents/sdk-go/blob/main/v2/event/extensions.go#L51
Im waiting for a response from them to see if its a bug or a feature :p

Reply:
the go sdk allows users to set mixed case and it fixes it for them in the same way the http header lib does.

Reply:
what happens when there is a name collision, for example `myExt` vs `myext`?

Reply:
@lionelvillard the validation in the Go SDK happens [inside `SetExtension()`](https://github.com/cloudevents/sdk-go/blob/82f2b61ecde41fd0577969cc58a7c2a18eeda250/v2/event/eventcontext_v1.go#L76-L81). At this point, `myExt` and `myext` can coexist like Scott mentioned.

When you call `Send()` via a CloudEvent client that uses the HTTP protocol, `SetExtension()` [overrides any existing header](https://github.com/cloudevents/sdk-go/blob/82f2b61ecde41fd0577969cc58a7c2a18eeda250/v2/protocol/http/write_request.go#L136) during the conversion. And because extensions are stored as a map, the header's value will be random.

Reply:
knative can choose to be more strict